### PR TITLE
trying to simplify current_act handling

### PR DESCRIPTION
**Pull request by ethn**

### CODE REVIEW

File: card/lib/card.rb [hunk 1]
```diff
@@ -120,7 +120,7 @@ class Card < ApplicationRecord
   has_many :drafts, -> { where(draft: true).order :id }, class_name: :Action
 
   cattr_accessor :set_patterns, :serializable_attributes, :error_codes,
-                 :set_specific_attributes, :current_act
```
Comment (ethn) [inline card/lib/card.rb]: now just use ActManager.act

File: card/lib/card/act_manager.rb [hunk 2]
```diff
@@ -154,17 +157,43 @@ def deep_delete director
...
+      
+      # If active jobs (and hence the integrate_with_delay events) don't run
+      # in a background process then Card::Env.deserialize! decouples the
+      # controller's params hash and the Card::Env's params hash with the
+      # effect that params changes in the CardController get lost
+      # (a crucial example are success params that are processed in
+      # CardController#update_params_for_success)
+      def contextualize act_id, card, env, auth
+        if delaying?
+          contextualize_for_delay(act_id, card, env, auth) { yield }
```
Comment (ethn) [inline card/lib/card/act_manager.rb]: rationale is that there are LOTS of things (env, auth, act, directors, clearing...) that need to happen if delayed but not if happening inline.  This harder separation cleans that up.

File: card/lib/card/act_manager.rb [hunk 3]
```diff
@@ -154,17 +157,43 @@ def deep_delete director
...
+        else
+          yield
+        end
+      end
+
+      def delaying?
+        Delayed::Worker.delay_jobs && Decko.config.active_job.queue_adapter == :delayed_job
```
Comment (ethn) [inline card/lib/card/act_manager.rb]: as requested.

File: card/lib/card/act_manager/stage_director/phases.rb [hunk 4]
```diff
@@ -26,9 +26,11 @@ def integration_phase
           run_single_stage :integrate_with_delay
         rescue => e # don't rollback
           Card::Error.current = e
-          warn "exception in integrate stage: #{e.message}"
-          warn e.backtrace.join "\n"
-          @card.notable_exception_raised
+          unless e.class == Card::Error::Abort
```
Comment (ethn) [inline card/lib/card/act_manager/stage_director/phases.rb]: I assume aborted actions are not notable errors.

File: card/lib/card/set/event/delayed_event.rb [hunk 5]
```diff
@@ -92,29 +71,34 @@ def serialize_value value
...
     when ActionController::Parameters
       serialize_value value.to_unsafe_h
     else
       { value: value }
     end
   end
 
+  def serialize_hash_value
+    value.each_with_object({}) { |(k, v), h| h[k] = serialize_value(v) }
+  end
```
Comment (ethn) [inline card/lib/card/set/event/delayed_event.rb]: no logical change intended.

File: card/mod/notifications/set/all/send_notifications.rb [hunk 6]
```diff
@@ -105,22 +105,17 @@ def current_act_card?
 end
 
 event :notify_followers_after_delete, :integrate,
-      on: :delete, when: proc { |ca| ca.notable_change? } do
+      on: :delete, when: proc { |ca|  ca.notable_change? } do
```
Comment (ethn) [inline card/mod/notifications/set/all/send_notifications.rb]: oops.

File: card/mod/notifications/set/all/send_notifications.rb [hunk 7]
```diff
@@ -105,22 +105,17 @@ def current_act_card?
...
-    if follower.account && follower != @current_act.actor
-      follower.account.send_change_notice @current_act, reason[:set_card].name,
-                                          reason[:option]
-    end
+  act = ActManager.act
+  act.reload
+  act_followers(act).each_follower_with_reason do |follower, reason|
+    next if !follower.account || (follower == act.actor)
+    follower.account.send_change_notice act, reason[:set_card].name, reason[:option]
   end
```
Comment (ethn) [inline card/mod/notifications/set/all/send_notifications.rb]: just wanted to break this out into two methods.

File: card/spec/lib/card/stage_director_spec.rb [hunk 8]
```diff
@@ -296,7 +296,7 @@ def event_called ev
             f:12
             f:1
             ig:1 ig:11 ig:111 ig:112v ig:12 ig:121
-            igwd:1 igwd:11 igwd:12 igwd:121 igwd:111 igwd:112v
+            igwd:1 igwd:11 igwd:111 igwd:112v igwd:12 igwd:121
```
Comment (ethn) [inline card/spec/lib/card/stage_director_spec.rb]: couldn't say why this happens, frankly.

File: card/lib/card/set/event/delayed_event.rb (hunk 5 above)
Comment (xithan) [reply card/lib/card/set/event/delayed_event.rb]: In this case you can change the logic to whatever you want. I don't think it is used anymore. I added a patch for ActiveJob that takes care of that (card/config/initializers/02_patches/active_job.rb)

File: card/mod/notifications/set/all/send_notifications.rb (hunk 6 above)
Comment (xithan) [reply card/mod/notifications/set/all/send_notifications.rb]: `when: :notable_change?` ?

File: card/lib/card/set/event/delayed_event.rb (hunk 5 above)
Comment (ethn) [reply card/lib/card/set/event/delayed_event.rb]: :)

**Pull request merged**